### DISCUSSION TITLE
Variant pairing

### DISCUSSION THREAD

**Discussion by nyyssont** [ORTools4MTEX/ORTools#76] [Q&A] [answered]
Hi,

the variant pairing histogram shows the variant pairs with respect to variant 1. Does it translate V2-VX to accommodate this sort of pairing or does the histogram only show those pairs where V1 really is a member?

-Tuomo

**Comment by nyyssont** [comment]
Maybe this is a better way to generate the V1-V24 list:

```
function [groupIds] = generateGroupIds()
% parent symmetry
cs_fcc = crystalSymmetry('m-3m', [3.6599 3.6599 3.6599], 'mineral', 'Iron fcc');

% child symmetry
cs_bcc = crystalSymmetry('m-3m', [2.866 2.866 2.866], 'mineral', 'Iron bcc');

% generate variants
varInt = variants(orientation.KurdjumovSachs(cs_fcc,cs_bcc),orientation.cube(cs_fcc,1));

% Generate misorientations between all variants
allMisos = inv(varInt)*varInt;

for l = 1:24
    % Compare the misorientations on each row to the first row (where
    % ordering is correct)
    [~,ind] = min(angle(allMisos(1,l),allMisos)/degree);
    % Make a list of variant Id pairs:
    groupIds_trial{l} = [1:24;ind]';
end

%Sort
groupIds_trial = cellfun(@(x) sort(x,2),groupIds_trial,'uniformOutput',false);
groupIds_trial = cellfun(@(x) sortrows(x),groupIds_trial,'uniformOutput',false);

%Remove duplicates
z = cell2mat(cellfun(@(x) x(:),groupIds_trial,'uniformOutput',false));
z = z';
z = unique(z,'rows','stable');
zz = num2cell(z,2);

%Back into cell
groupIds_trial = cellfun(@(x) reshape(x,[],2),zz,'uniformOutput',false);

%Remove duplicate rows, ready to be fed to computeGrainPairs
groupIds = cellfun(@(x) unique(x,'rows'),groupIds_trial,'uniformOutput',false);
groupIds = groupIds(2:end);
end
```

**Reply by nyyssont** [reply]
After some fiddling, I think it works. The result generated is close but not quite equivalent to when you use the list in Example 10:
```
vGroupIds = {[1 2],...
    [1 3; 1 5],...
    [1 4],...
    [1 6],...
    [1 7],...
    [1 8],...
    [1 9; 1 19],...
    [1 10; 1 14],...
    [1 11; 1 13],...
    [1 12; 1 20],...
    [1 15; 1 23],...
    [1 16],...
    [1 17],...
    [1 18; 1 22],...
    [1 21],...
    [1 24]};
```

![image](https://github.com/user-attachments/assets/b96094a1-f65f-4524-b9f2-9f280c5e66a5)

**Reply by frankNiessen** [reply]
Hi Tuomo, 

sorry for the late reply, I am rather busy these days ... 

So I can tell you that ORTools by default returns all the variant pairs: 
`computeGrainPairs(grains,'variants','plot');
`![image](https://github.com/user-attachments/assets/656f9458-2cbb-49fd-8ef6-9443cbc9b00a)

If you want a 1D histogram you will have to make a decision what you want when defining the groups. In a scenario with 6 variants, the grouping
`vGroupIds = {[1,1],[1 2],[1,3],[1,4],[1,5],[1,6]}` would correspond to your last sugestion: Only the 1-X pairs, no equivalency. 

`computeGrainPairs(grains,'variant','group',vGroupIds,'plot');
`I think what you suggest is what you want in most cases - finding all equivalent misorientations with respect to the V1-Vx misorientation and plot these. I would in your code check for a matching misorientation axis as well. 
You should achieve this by defining `vGroupIds` according to your `groupIds_trial` variable.

I will see if I can get Azdiar to comment as well, he knows the implementation better than I do.

**Comment by frankNiessen** [comment, marked as answer]
Hey @nyyssont , 

I am sorry that we fell short in doing something about this earlier, but alas, the fix is in with commit
https://github.com/ORTools4MTEX/ORTools/commit/e4e23b96ba8064aaf24688af6ea5bbf1414ba21c .

You were right that we were discarding the majority of the data by only considering the boundaries that actually had a V1-VX pairing and discarding all other equivalent boundaries. We now take all 276 variant pairings into account. Your approach went into a new function `computeVariantGroups`.

Running example 10 pre and post commit https://github.com/ORTools4MTEX/ORTools/commit/e4e23b96ba8064aaf24688af6ea5bbf1414ba21c does not show a huge difference in the histogram, but the statistical backing is about an order of magnitude larger.
 

BEFORE BUGFIX
<img width="521" height="407" alt="image" src="https://github.com/user-attachments/assets/e34a1c93-e739-479c-8cef-bcf44496420f" />


AFTER BUGFIX
<img width="521" height="407" alt="image" src="https://github.com/user-attachments/assets/537e7172-f01d-4544-93ae-eebc238e56fb" />
